### ISSUE TITLE
buildLayeredImage has surprising behaviour wrt. symlink paths on different backends

### ISSUE THREAD

**Issue by srhb** [NixOS/nixpkgs#107021]
**Describe the bug**
Please see https://github.com/NixOS/nixpkgs/issues/107021#issuecomment-753984110 for a more thorough investigation than this.

~I'm not sure I'd really classify this as a bug, but it is a least surprising behaviour on some docker storage backends, so I'd like to document this. I'm not sure whether docker really has any formal semantics with respect to the loading process.~

The TL;DR: Some `buildLayeredImage`s may be usable on some storage backends but not on others when symlinked paths are involved.

**To Reproduce**

Consider:

```
let
  nixpkgs = builtins.fetchTarball "https://github.com/NixOS/nixpkgs/archive/e9158eca70ae59e73fae23be5d13d3fa0cfc78b4.tar.gz";
  pkgs = import nixpkgs { config = {}; overlays = []; };

  real = pkgs.writeTextDir "dir/real" "contents";
  symlink = pkgs.runCommandNoCC "symlink" {} ''
    ln -s ${real} $out
  '';
in
pkgs.dockerTools.buildLayeredImage {
  name = "symlink-image";
  contents = [symlink];
}
```

This image fails to load on an ext4/overlayfs setup:

```
49ebdedff0c8: Loading layer [==================================================>]  10.24kB/10.24kB
e12f90c968f3: Loading layer [==================================================>]  10.24kB/10.24kB
Error processing tar file(exit status 1): mkdir /nix/store/58f4852rq2517gayjdc8lnvr6chklsgc-symlink/dir: no such file or directory
```

But succeeds on a zfs-backed docker setup:
```
49ebdedff0c8: Loading layer [==================================================>]  10.24kB/10.24kB
e12f90c968f3: Loading layer [==================================================>]  10.24kB/10.24kB
20125547c15a: Loading layer [==================================================>]  10.24kB/10.24kB
Loaded image: symlink-image:yf3llawsrlqky9cnmpv29hkyys368329
```

I imagine this is because of the storage backend's behaviour wrt. earlier layers' files.

**Expected behavior**
Image loads the same on all backends.

**Additional context**

This was discovered because our docker images became unpullable on some hosts (overlay) but not others (zfs) after crate2nix started using symlinks for tested builds: https://github.com/kolloch/crate2nix/pull/143

@andir might find this interesting. :smile_cat:

**Comment by andir** [comment]
🤣 docker.. sounds like an upstream issue, no? What happens if you do the
equivalent in a Dockerfile?

**Comment by srhb** [comment]
> docker.. sounds like an upstream issue, no?

Yes, I tend to agree, both backends should behave the same, but I'm afraid "overlay" is the de facto standard, so the functioning version is probably the buggy one. I may report this upstream, but it's at least surprising that `buildLayeredImage` can produce unpullable images (never happened to me before!) :grinning: 

> What happens if you do the equivalent in a Dockerfile?

Uh, that would require a working memory on how to make Dockerfiles... Stand by. :stuck_out_tongue:

**Comment by srhb** [comment]
I can't seem to reproduce this with Dockerfiles. I suspect it does some logic to avoid having symlink breakage across layers, but I haven't checked the source. Maybe we should do the same.

**Comment by srhb** [comment]
Scratch the above, my assumptions were wrong. Here's what I now think is going on, seen only from the perspective of an overlayfs setup (forget about zfs, I've no idea what process the clone/snapshot mechanism uses to make this magically "work" and to what extent it actually "works")

Given the snippet:

```
let
  # Bad
  nixpkgs = builtins.fetchTarball "https://github.com/NixOS/nixpkgs/archive/560201da661a4c916d0732e3bab8a77084485659.tar.gz";

  # Good
  # nixpkgs = builtins.fetchTarball "https://github.com/NixOS/nixpkgs/archive/560201da661a4c916d0732e3bab8a77084485659~1.tar.gz";

  pkgs = import nixpkgs { config = {}; overlays = []; };

  real = pkgs.writeTextDir "dir/real" "contents";
  symlink = pkgs.runCommandNoCC "symlink" {} ''
    ln -s ${real} $out
  '';
in
pkgs.dockerTools.buildLayeredImage {
  name = "symlink-image";
  contents = [symlink];
}
```

We get an image which has a layer  `39fff2240b0ade19727688458df91f31732033aae12885daeb34064a8b4e3ac2/layer.tar` containing:

```
 drw-r--r--       0/0             0 /nix/
 drw-r--r--       0/0             0 /nix/store/
 lrwxrwxrwx  nobody/nogroup       0 /nix/store/qh48dhi2rgxhrg26pssvgcxc2mvw1ddk-symlink --> /nix/store/v1v4cb7xyplfhqkk7jnfwh5y7h1r4x5r-real
 dr-xr-xr-x  nobody/nogroup       0 /nix/store/qh48dhi2rgxhrg26pssvgcxc2mvw1ddk-symlink/dir/
 -r--r--r--  nobody/nogroup       8 /nix/store/qh48dhi2rgxhrg26pssvgcxc2mvw1ddk-symlink/dir/real
```

This, I think, is the main culprit. If we go to the prior commit (the commented nixpkgs revision) the problem does not exist, and the corresponding layer looks like this:

```
 lrwxrwxrwx    root/root          0 /nix/store/qh48dhi2rgxhrg26pssvgcxc2mvw1ddk-symlink --> /nix/store/v1v4cb7xyplfhqkk7jnfwh5y7h1r4x5r-real
```

Which looks much more correct and doesn't have more than one entry for `/nix/store/qh48dhi2rgxhrg26pssvgcxc2mvw1ddk-symlink` with one being a symlink (correct) and the other (implicitly) a directory.

In other words, I believe this is a bug and was introduced in #91084 specifically in 560201da661a4c916d0732e3bab8a77084485659 and relates to how disparate entries are somehow created because of some wonky symlink traversal. :smile_cat: 

cc @utdemir @purcell
